Latency aware policy

wprzytula opened the pull request:
A LatencyAware load balancing policy, similar to what other drivers deliver, was added. Its goal is to avoid slow nodes as coordinators.
If a policy requests that latency statistics be collected, then an exponential moving average of latencies will be updated for a node upon query completion with that node serving as coordinator.

Fixes: #515 #505 #260

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented:
Tests have yet to be done.

wprzytula commented:
Tests are added, ready for review: @psarna @piodul @cvybhu

piodul commented on scylla/src/transport/load_balancing/latency_aware.rs:
```diff
@@ -0,0 +1,253 @@
+    }
+
+    fn fast_enough(
+        node: &Arc<Node>,
```
Nit: you are not cloning the `&Arc<Node>`, so you can as well change the argument type to `&Node` so that this function is able to accept a wider range of arguments. Besides, `&Arc<_>` is a pointer-to-a-pointer, so it may be that the program needs to do double indirection instead of one in order to access the contents (unless the optimizer does its job).

piodul commented on scylla/src/transport/session.rs:
```diff
@@ -1162,10 +1162,10 @@ impl Session {
-                    QueryError::TooManyOrphanedStreamIds(_) => todo!(),
-                    QueryError::UnableToAllocStreamId => todo!(),
-                    QueryError::DbError(DbError::IsBootstrapping, _)
+                    QueryError::BadQuery(_)
+                    | QueryError::TooManyOrphanedStreamIds(_)
+                    | QueryError::UnableToAllocStreamId
+                    | QueryError::DbError(DbError::IsBootstrapping, _)
```
This fix belongs to some previous commit, probably.

piodul commented on scylla/src/transport/load_balancing/latency_aware.rs:
```diff
@@ -251,3 +251,332 @@ impl<'a> Iterator for IteratorWithSkippedNodes<'a> {
+            ],
+        );
+
+        tokio::time::sleep(policy.retry_period).await;
```
I'd recommend sleeping a bit more (e.g. twice the amount of milliseconds) than the retry period - the tokio timer's precision isn't perfect.

wprzytula replied:
On the other hand, retry period is by default measured in seconds, so its precision should be good enough. Should we bother with much smaller retry period configurations?

cvybhu commented on scylla/src/transport/session.rs:
```diff
@@ -1321,12 +1351,16 @@ impl Session {
 
+                let elapsed = query_start.elapsed();
+                if self.should_consider_query_for_latency_measurements(&query_result) {
+                    let mut average_latency_guard = node.average_latency.write().unwrap();
```
We could create a channel on which latency measurements would be sent.
`NodeWorker` would collect these measurements and periodically update atomic with current latency.

This would offload the cost of doing `thread_local` super optimized stuff onto the channel library,  something like crossbeam.

cvybhu replied:
We could create a channel on which latency measurements would be sent.
`NodeWorker` would collect these measurements and periodically update atomic with current latency.

This would offload the cost of doing `thread_local` super optimized stuff onto the channel library,  something like crossbeam.

wprzytula replied:
I consider this problem as open, and I have no personal preference nor conviction towards any particular solution. It seems though to be the last (yet important) issue with the PR, so I encourage you to further discussion.

piodul replied:
I'm not sure if I like the channel solution because either the channel has to be bounded and we might have to either drop measurements/block until the channel is cleared, or the channel is unbounded and its memory usage is theoretically unbounded.

IMO for now we can proceed with the current solution - if the original implementation performs similar locks then we shouldn't be much worse performance wise. In the future, if we have more data and the locks turn out to be a bottleneck, we might consider optimizing it. Ultimately, all those problems are opt-in and require the latency-aware policy to be turned on.

@psarna @cvybhu what do you think?

psarna replied:
Let's release fast and optimize on demand.

cvybhu commented on scylla/src/transport/load_balancing/token_and_latency_aware.rs:
```diff
@@ -0,0 +1,278 @@
+impl Default for TokenAndLatencyAwarePolicy {
+    fn default() -> Self {
+        Self {
+            inner: TokenAwarePolicy::new(Box::new(LatencyAwarePolicy::default())),
```
I think the hierarchy should be"
`LatencyAwarePolicy` -> `TokenAwarePolicy` -> `RoundRobin`

`TokenAwarePolicy` would choose replicas and shuffle them with round robin, then LatencyAware would use this plan, but filter out nodes that have high latencies.

This doesn't really fit the `ChildLoadBalancingPolicy` trait, but we could make `LatencyAwarePolicy` keep `inner: Box<dyn LoadBalancingPolicy>` inside of it.

psarna replied:
> * `LatencyAwarePolicy` will only be applied to replicas, after trying all the replicas other nodes will be tried without regard for latency

I think this is what we want. Ultimately we're going to contact replicas anyway, so we should prefer to do it directly. There are cases where contacting a non-replica coordinator is better because replicas are overloaded, but in my opinion it's far more likely that contacting a replica is more beneficial in the average case.

> * There is no load balancing among replicas. Each query with a token for replica `A` will always hit `A` first, then `B` and `C`. With round robin the load is spread equally among the replicas, because the first replica changes.

Could you clarify your concerns? Assuming replication factor == 3, each token would determine a set of 3 replicas that own it. It sounds correct to apply latency awareness on this filtered list of candidates.

Also, take into account that we should generally assume that gathering latency information is a costly operation (it's prohibitively costly right now with a mutex, but it will still be costly on atomics). Therefore, getting this information from a pre-filtered list of replicas sounds more efficient than gathering latency information on all nodes (imagine we have tens of nodes in multiple datacenters) and only then applying token awareness. I think the current approach is valid.

cvybhu replied:
My main problem with this implementation is lack of load balancing among the replicas. There is no round robin, so for a given token we will always try replicas in the same order. This will make the first replica handle all the requests, while the others stay idle.
I understand that this is a `LatencyTokenPolicy`, not `LatencyTokenRoundRobinPolicy`, but I think that we should introduce a policy that uses the latency information and is usable by production code.

cvybhu replied:
> I think this is what we want. Ultimately we're going to contact replicas anyway, so we should prefer to do it directly. There are cases where contacting a non-replica coordinator is better because replicas are overloaded, but in my opinion it's far more likely that contacting a replica is more beneficial in the average 

You're right. Still after trying the replicas it would be good to try other nodes in the order of their latencies, but I'm not sure if there's an easy way to do it.

psarna commented on scylla/src/transport/session.rs:
```diff
@@ -1151,6 +1153,34 @@ impl Session {
+                    | QueryError::DbError(DbError::IsBootstrapping, _)
+                    | QueryError::DbError(DbError::Unavailable { .. }, _)
+                    | QueryError::DbError(DbError::Unprepared { .. }, _)
+                    | QueryError::DbError(DbError::Overloaded { .. }, _) => false,
```
That's an interesting problem. `Overloaded` error actually means "coordinator is overloaded" in Cassandra terminology, and in fact we'd receive this error rather quickly, because from the database server perspective, the sooner we disqualify the request, the better, because then we don't waste resources on a query that's doomed to fail. So, on the one hand, it's deceptive to count an `Overloaded` response into the latency tracker, because this response may come very fast, even though the coordinator is indeed overloaded. Drivers should be as cooperative as possible, so perhaps we should have a special case for `Overloaded` error, and induce fake high latency to the tracker (e.g. 10s), to make it less likely to pick this node in the near future. @piodul I'd appreciate your take on this too

piodul replied:
I agree that it doesn't make sense to count the actual latency of a request that resulted in `Overloaded` response.

I'm not sure whether I completely understand the behavior of the policy, so I can't tell if 10 seconds in particular (or any other constant) will be all right for all cases. Instead, in case we get the `Overloaded` response, we can set the node's latency to `max(current_latency, 2 * min_latency)` so that the node is naturally retried after the configured `retry_period` but the latency doesn't grow too much and there isn't too much catching up needed for the moving average to do.

piodul commented on scylla/src/transport/node.rs:
```diff
@@ -10,17 +10,56 @@ use std::{
+            }),
+            Some(prev_avg) => Some({
+                let delay = (now - prev_avg.timestamp).as_nanos() as f64;
+                let prev_weight = (delay + 1.).ln() / delay;
```
Shouldn't `delay` be expressed in seconds? IIUC the original code gets the number of elapsed nanoseconds but then divides it by `1e9` (`scaled_delay`) and, finally, calculates `prev_weight` using seconds in the formula.

wprzytula commented on scylla/src/transport/session.rs:
```diff
@@ -1151,6 +1153,34 @@ impl Session {
+                    QueryError::DbError(_, _)
+                    | QueryError::InvalidMessage(_)
+                    | QueryError::IoError(_)
+                    | QueryError::ProtocolError(_)
```
I followed Java driver implementation in that case. Should we conform or not?

psarna replied:
It's fine to deviate if we're confident our decision is better. I think that in this particular case we can leave the code as is (perhaps the Java devs had their own reasons supported by tests that it's a good idea), and revise our decision if anybody reports a problem.

wprzytula commented:
v2:
- fixed minor nits,
- reworked the policy so that last min avg is updated periodically by a worker, and is contained in an atomic instead of a lock,
- fixed delay unit (nanos to secs),
- rebased on current main.

wprzytula commented on scylla/src/transport/load_balancing/latency_aware.rs:
```diff
@@ -0,0 +1,295 @@
+                let nodes = nodes_rx.latest();
+                if let Some(nodes) = nodes {
+                    if nodes.is_empty() {
+                        error!("Empty node list - driver/policy bug!");
```
I no requests are sent for some time, then the channel will remain empty, 

> I can see that the nodes channel starts being empty. What if, after a session is created with the latency aware policy, no requests are being sent for some time? Won't this cause the loop to break?

If no requests are sent for some time, then the channel will remain empty and the `if let` block won't be entered, so the loop will be silently continued.

wprzytula commented on scylla/src/transport/load_balancing/latency_aware.rs:
```diff
@@ -0,0 +1,295 @@
+
+impl LoadBalancingPolicy for LatencyAwarePolicy {
+    fn plan<'a>(&self, _statement: &Statement, cluster: &'a ClusterData) -> Plan<'a> {
+        self.make_plan(cluster.all_nodes.clone())
```
IIUC, Java driver's implementation behaves the same: it does not shuffle the replicas on its own. The idea is that one should use a round robin flavour of this policy for shifting the "fast" and the "slow" replica sets with each query. Perhaps it is a good idea to add such a policy, RoundRobinLatencyAware policy? And maybe even RoundRobinLatencyAndTokenAware policy?

piodul replied:
I'd rather not add every non-empty combination of {TokenAware, LatencyAware, DCAware, RoundRobin} to the driver. We can take one of the two approaches to combat the exponential grow:

- Implement the LatencyAwarePolicy as a "wrapper" policy. It will contain a child policy and delegate some details to it using the existing ChildLoadBalancingPolicy trait. We already to it for TokenAware policy and DCAwarePolicy, I think. On the other hand, it's a bit hard to think about how the policy will actually behave when composed. We are moving away from this approach in the load balancing policy refactor which was worked on by @havaker / @cvybhu (https://github.com/scylladb/scylla-rust-driver/pull/491).
- Implement a LatencyAwarePolicy which supports all of the features (token-awareness, DC awareness, round robin) but allows to switch them off in runtime with boolean flags.

I'm not sure which approach is better. @cvybhu thoughts?

wprzytula replied:
I've implemented the first approach. Looking forward for reviews. @piodul @cvybhu

wprzytula commented on scylla/src/transport/load_balancing/latency_aware.rs:
```diff
@@ -0,0 +1,295 @@
+        &self,
+        plan: Vec<Arc<Node>>,
+    ) -> Box<dyn Iterator<Item = Arc<Node>> + Send + Sync> {
+        self.make_plan(plan)
```
That's unfortunately true.
However, let's discuss what problems might arise from that. The direct consequence is that `min_avg` can be different that it should (more specifically, it can only be **greater** that it should be, due to the properties of `min` function called on a subset instead of the whole set). If `min_avg` is greater than it should be, some nodes may not be penalised while they should, and this can alter the desired order of querying. An example:
Node | avg latency (ms)
A       |  10
B       |  100
C       |  150
Let's assume that A was excluded due to token awareness, and so `min_avg` updater computed `100 ms` as the min_avg. If we now make a query that in aspect of token awareness is present on A and B, and moreover B is before A on the initial nodes list, then B won't be penalised and it will become the coordinator of the query, even though A is a much better candidate for that due to 10x smaller latency.

piodul commented on scylla/src/transport/load_balancing/token_and_latency_aware.rs:
```diff
@@ -0,0 +1,305 @@
+
+/// A token- and latency-aware load balancing policy.
+#[derive(Debug)]
+pub struct TokenAndLatencyAwarePolicy {
```
I guess this should be removed now? I think it is possible now to compose `LatencyAwarePolicy` with `TokenAwarePolicy` manually.

piodul replied:
The tests might be useful though in order to verify that the policies compose in a sane way.

wprzytula commented on scylla/src/transport/load_balancing/latency_aware.rs:
```diff
@@ -0,0 +1,704 @@
+        let min_latency = Arc::new(AtomicDuration::new());
+        let mut update_scheduler = tokio::time::interval(update_rate);
+        let ask_for_nodes = Arc::new(AtomicBool::new(true));
+        let (mut nodes_rx, nodes_tx) = single_value_channel::channel::<Vec<Arc<Node>>>();
```
The single value channel has API that is distinct from bare `Arc<Mutex<Option<T>>>` in that it clearly separates writing and reading sides. Of course we can refrain from adding the dependency, but its purpose is valid here.

wprzytula replied:
After all, the whole purpose of using types is reducing the possible actions to performs on data to those which are purpose- and meaning- -ful

piodul replied:
I understand, but IMO it's not worth it. The usage of `nodes` is simple enough here that typing `nodes.lock().unwrap().clone()` instead of `nodes.latest()` doesn't obscure the algorithm too much, and it's not hard to track who is the writer and who is the reader. On the other hand, seeing a non-standard crate like this makes you wonder how it works and you have to learn a new abstraction.

wprzytula commented on scylla/src/transport/load_balancing/latency_aware.rs:
```diff
@@ -0,0 +1,704 @@
+    }
+
+    fn refresh_last_min_avg_nodes(&self, nodes: &[Arc<Node>]) {
+        if self.ask_for_nodes.swap(false, Ordering::Relaxed) {
```
> The flag will be false most of the time so, as an optimization, I'd recommend performing a relaxed atomic load before attempting to swap the flag.
I can't understand what I should load and how it is supposed to work.

piodul replied:
```rust
fn refresh_last_min_avg_nodes(&self, nodes: &[Arc<Node>]) {
    if self.ask_for_nodes.load(Ordering::Relaxed) // <- this is cheaper than swap and will be false most of the time
            && self.ask_for_nodes.swap(false, Ordering::Relaxed) {
        let _ = self.nodes_tx.update(Some(nodes.to_owned()));
        trace!("LatencyAwarePolicy: updating nodes list for the min average updater");
    }
}
```

piodul commented on scylla/src/transport/session.rs:
```diff
@@ -1298,6 +1328,7 @@ impl Session {
 
         let runner = async {
             let cluster_data = self.cluster.get_data();
+            self.load_balancer.update_cluster_data(&cluster_data);
```
You need to add a similar line in `iterator.rs` as well. It can be placed somewhere at the beginning of the `RowIteratorWorker::work()` function.

As a side note, we should probably try to think about refactoring `RowIteratorWorker::work` and `Session::execute_query` into common code... They are a bit similar and also complicated at the same time, and it's easy to miss one of them when modifying the other. OFC, it's not something that should be done in this PR.

piodul replied:
Latency measurements should be updated in `iterator.rs` if possible, too.

wprzytula commented:
v3: further requested fixes (update_cluster_data for child policy, updating measurements in iterator.rs)

wprzytula commented:
We've benchmarked the new policy in order to assess its performance. The results are truly disappointing, as it appears to be worse than the default policy in every case, even one that I should thrive in. If we decide to merge such a policy anyway, I strongly suggest documenting it with firm declarations that the policy, according to our benchmarks, works simply bad.

**Main observations**:
- latency is twice as bad (comparing to the default policy) in a scenario that one node is heavily loaded. 
- latency is worse also in neutral situation (all nodes equally loaded).



"Default policy" is TokenAware RoundRobin.

**With RF=1** (no point in Latency Aware Policy here):

Default policy:
![Throughput_RF=1_Default_Policy](https://user-images.githubusercontent.com/59482568/196475209-19885c2a-36f4-4eda-9eb4-f00fd88b296f.png)
![Latency_RF=1_Default_Policy](https://user-images.githubusercontent.com/59482568/196475242-f2d86f3d-8ed9-4ffc-a05e-a781050d038f.png)
![Requests_per_coordinator_RF=1_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196475265-bd8ca1a9-43f2-4841-a7dc-09eee6921de5.png)
![Writes_per_replica_RF=1_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196475274-5cfb28d6-8ed9-4f2c-a517-79b5aec16cd6.png)


LatencyAware policy:

![Throughput_RF=1_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196475490-405c870b-2bd3-4305-bfa4-0793d4601b83.png)
![Latency_RF=1_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196475503-e8d24449-51f2-4b1c-a410-886408e4300f.png)
![Requests_per_coordinator_RF=1_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196475513-804408e4-1db7-402d-998e-f0b5b90add3c.png)
![Writes_per_replica_RF=1_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196475522-7ab9b19e-4fe4-4644-9a07-2f00998111fd.png)

**With RF=3** (no disruptions):
Default Policy:
![Throughput_RF=3_Default_Policy](https://user-images.githubusercontent.com/59482568/196475693-3307bd1e-72f9-4462-abef-8fbb6cd9d225.png)
![Latency_RF=3_Default_Policy](https://user-images.githubusercontent.com/59482568/196475710-459fedca-80bc-4c57-9a41-c810f0cc65ad.png)
![Requests_per_coordinator_RF=3_Default_Policy](https://user-images.githubusercontent.com/59482568/196475764-5f83e8e7-c997-4de5-b184-63be04e98906.png)
![Writes_per_replica_RF=3_Default_Policy](https://user-images.githubusercontent.com/59482568/196475774-886ad793-bded-4a35-b942-6dd43b5de291.png)

LatencyAwarePolicy:
![Throughput_RF=3_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196475877-08e5e5d7-d388-4c17-a312-282ab3c6c2a2.png)
![Latency_RF=3_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196475891-d9ef958c-8868-4baa-a0f0-79c89194e7eb.png)
![Requests_per_coordinator_RF=3_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196475903-62cc3c07-03d7-4702-b78b-f33b8f4f7f8b.png)
![Writes_per_replica_RF=3_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196475918-79ab90fc-9b8e-4158-8a96-aaae7be7da54.png)

**With RF=3**, one node heavily loaded with some CPU-intensive computations:

Default policy:
![Throughput_RF=3_one_slow_node_Default_Policy](https://user-images.githubusercontent.com/59482568/196476135-98994e71-ef60-4630-b250-efb6e99be25c.png)
![Latency_RF=3_one_slow_node_Default_Policy](https://user-images.githubusercontent.com/59482568/196476155-a371606b-f12a-4865-b0ad-a5b38d24586d.png)
![Requests_per_coordinator_RF=3_one_slow_node_Default_Policy](https://user-images.githubusercontent.com/59482568/196476181-213b26e9-5b3d-4bcd-8a91-192917c46b27.png)
![Writes_per_replica_RF=3_one_slow_node_Default_Policy](https://user-images.githubusercontent.com/59482568/196476195-ff5face8-6fae-46cf-b8e5-77cc37ef74d1.png)
![Load_RF=3_one_slow_node_Default_Policy](https://user-images.githubusercontent.com/59482568/196476355-05b9f5cd-6d88-4317-90b8-b1fc96974c25.png)
![Foreground_writes_RF=3_one_slow_node_Default_Policy](https://user-images.githubusercontent.com/59482568/196476364-51c10fc3-fc6a-44d2-aeb5-00f62247e4d2.png)


LatencyAware policy:
![Throughput_RF=3_one_slow_node_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196476416-57aa2cfe-834a-4009-bbe9-de20c1cce5dd.png)
![Latency_RF=3_one_slow_node_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196476427-e15872f3-4fa4-4093-91d4-152a8fdb5fb6.png)
![Requests_per_coordinator_RF=3_one_slow_node_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196476447-6f52def7-d907-4d5c-b786-662d1cf204ba.png)
![Writes_per_replica_RF=3_one_slow_node_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196476458-631e58d2-645a-456e-86fc-4709d630e651.png)
![Load_RF=3_one_slow_node_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196476483-63650a60-0ef3-46ee-a047-e8a008e10288.png)
![Foreground_writes_RF=3_one_slow_node_Latency_Aware_Policy](https://user-images.githubusercontent.com/59482568/196476493-00ad89f4-94f1-439c-9d71-d682398a3c22.png)

piodul commented:
I'm not sure if the results are that alarming. Throughput looks comparable in all cases. As for latency, the baseline latency looks comparable, too; because this is a write-only benchmark, there will be memtable flushes and compactions which might schedule themselves differently in different runs and have different effect on p95 and p99 shown in the graphs.

Apart from that, I'm not sure if a write-only workload is a good benchmark for the latency aware policy. In Scylla, write coordinators actually wait in the background for all replicas to acknowledge writes, and will slow down further writes if there are too many unacknowledged writes in the background, effectively slowing down to the speed of the slowest replica. If latency is bad on one node, then it should also be bad on other nodes.

As a side note, in the benchmark with one node being overloaded it doesn't look like the overloaded node is penalized. Either penalizing does not work in the policy, or the latency is bad due to all nodes being affected by the overload due to the reason I mentioned in the previous paragraph (I suspect the latter).

A read-only workload would be a better choice because reads don't contact all replicas - for example, if you used CL=QUORUM then only two replicas will be contacted. Here, you should observe some gains compared to the default policy. I suggest redoing the benchmark in read-only mode.

Finally, you are showing server-side latency, while it would be best to measure the latency on the client side - because it's what affects users the most. I'm not sure which tool you are using for benchmarks, but I can recommend my cql-stress, which measures latency on the client side: https://github.com/piodul/cql-stress/
